engines/aws/app/controllers/aws/credentials_controller.rb

```diff
@@ -6,14 +6,16 @@ def edit; end
 
     def update
       @credential = Credential.new(self.credential_params)
-      if @credential.save
+      valid_credentials = @credential.valid_credentials? self.credential_params
+      if valid_credentials && @credential.save
         flash[:success] = t(
           'engines.aws.credentials.using',
           key_id: @credential.aws_access_key_id
         )
         redirect_to(helpers.next_step_path(aws.edit_credential_path))
       else
-        flash[:warning] = @credential.errors.full_messages
+        flash[:warning] = @credential.errors.full_messages if valid_credentials
```

Comment:
Formatting question: This is two clauses in one? Probabaly normal in Ruby syntax but from Python standpoint I would split everything out:

```
if not valid_credentials
  flash[:warning] = "Credentials are not valid."
elif not @credential.save
  flash[:warning] = @credential.errors.full_messages
else
  flash[:success] = t(
    'engines.aws.credentials.using',
    key_id: @credential.aws_access_key_id
  )
  redirect_to(helpers.next_step_path(aws.edit_credential_path))
end
```